File: frontend/src/domain/audio/value-objects/time-range.ts
/**
 * Time range value object.
 */

export class TimeRange {
    constructor(
        public readonly start: number,
        public readonly end: number
    ) {
        if (start < 0) {
            throw new Error('Start time cannot be negative');
        }
        if (end < 0) {
            throw new Error('End time cannot be negative');
        }
        if (start >= end) {
            throw new Error('Start time must be before end time');
        }
    }

    static create(start: number, end: number): TimeRange {
        return new TimeRange(start, end);
    }

    get duration(): number {
        return this.end - this.start;
    }

    contains(time: number): boolean {
        return time >= this.start && time <= this.end;
    }

    overlaps(other: TimeRange): boolean {
        return this.start < other.end && this.end > other.start;
    }

    equals(other: TimeRange): boolean {
        return this.start === other.start && this.end === other.end;
    }
}

File: backend/streamcraft/domain/audio/value_objects/time_range.py
"""Time range value object."""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class TimeRange:
    """Represents a time range with start and end."""

    start: float  # seconds
    end: float  # seconds

    def __post_init__(self) -> None:
        """Validate time range."""
        if self.start < 0:
            raise ValueError("Start time cannot be negative")
        if self.end < 0:
            raise ValueError("End time cannot be negative")
        if self.start >= self.end:
            raise ValueError("Start time must be before end time")

    @property
    def duration(self) -> float:
        """Get duration in seconds."""
        return self.end - self.start

    def contains(self, time: float) -> bool:
        """Check if time is within range."""
        return self.start <= time <= self.end

    def overlaps(self, other: "TimeRange") -> bool:
        """Check if this range overlaps with another."""
        return self.start < other.end and self.end > other.start
